### PR TITLE
Implement slack notifier

### PR DESCRIPTION
**Pull request by elioschmutz**
Main implementation of the `ftw.slacker`.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,21 +1,278 @@
...
+===========
+
+.. contents:: Table of Contents
+
+
 Introduction
-============
+------------
+
+The `ftw.slacker` provides an easy to use api to post messages into a Slack channel through webhooks in Plone.
```
Comment (jone) [inline README.rst]: `` `ftw.slacker` `` => ``` ``ftw.slacker`` ```

Comment (jone) [inline README.rst]: The `in Plone` is confusing me a bit.

Maybe:
```diff
- The `ftw.slacker` provides an easy to use api to post messages into a Slack channel through webhooks in Plone.
+ The `ftw.slacker` is a Plone addon that provides an easy to use api to post messages into a Slack channel through Slack's webhooks api.
```
- I think it is important to declare that we are using the slack webhooks API (that's the difference to`slacker` for example); therefore `through Slack's webhooks api`.
- It was unclear what the role of Plone is; therefore I moved it to the front and declared that `ftw.slacker` is a `Plone addon`.

File: README.rst [hunk 2]
```diff
@@ -1,21 +1,278 @@
...
+
+For more information about slack webhooks see Slacks documentation about `Incoming Webhooks <https://api.slack.com/incoming-webhooks>`_
```
Comment (jone) [inline README.rst]: `Slacks` => `Slack's`

File: README.rst [hunk 3]
```diff
@@ -1,21 +1,278 @@
...
+
+First of all, you need to setup a Slack webhook.
+
+Read the Slack documentation about `Incoming Webhooks <https://api.slack.com/incoming-webhooks>`_ and start
+setting up your own webhock by follow the `incoming webhook integration <https://my.slack.com/services/new/incoming-webhook/>`_.
+
+Post message to Slack
+~~~~~~~~~~~~~~~~~~~~~
+
+Just import the `notify_slack` api method and call it.
```
Comment (jone) [inline README.rst]: `method` => `function`

File: README.rst [hunk 4]
```diff
@@ -1,21 +1,278 @@
...
+Configure the requests module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Following properties are passed to the requests module:
+
+- `webhook_url <http://docs.python-requests.org/en/master/user/quickstart/>`_
+- `timeout <http://docs.python-requests.org/en/master/user/quickstart/#timeouts>`_
+- `verify <http://docs.python-requests.org/en/master/user/advanced/#ssl-cert-verification>`_
+
+That means, you can call the api-method with this parameteres to configure the request:
```
Comment (jone) [inline README.rst]:
```diff
- That means, you can call the api-method with this parameteres to configure the request:
+ That means you can call the api-function with these parameters to configure the request:
```

File: README.rst [hunk 5]
```diff
@@ -1,21 +1,278 @@
...
+
+.. code:: python
+
+  from ftw.slacker import notify_slack
+
+  notify_slack(webhook_url='https://hooks.slack.com/services/xxx',
+               timemout=10,
+               verify=False,
+               text="my first post")
+
```
Comment (jone) [inline README.rst:83]: It would be nice to have a statement about the return value. I suggest:
```
The ``notify_slack`` function does not return a succes state nor raise an exception when an error occurs,
beacuse the message is posted in another thread which may take longer to finish.
```

File: README.rst [hunk 6]
```diff
@@ -1,21 +1,278 @@
...
+Slack payload
+~~~~~~~~~~~~~
+
+Just add additional keyword arguments to the api-method. All parameters will be passed
```
Comment (jone) [inline README.rst]: `method`= > `function` 😉

File: README.rst [hunk 7]
```diff
@@ -1,21 +1,278 @@
...
+               text="my first post",
+               attachments=[
+                   {
+                       "title": "Slack API Documentation",
+                       "title_link": "https://api.slack.com/",
+                       "text": "Optional text that appears within the attachment"
+                   }
+               ])
+
+Webhook URL by envrionment variables
```
Comment (jone) [inline README.rst]: `envrionment` => `environment`

File: README.rst [hunk 8]
```diff
@@ -1,21 +1,278 @@
...
+------------------------------------
+
+Normally you don't want to store your webhook-url in your application code.
+
+The api-method provides environment-variables configuration:
```
Comment (jone) [inline README.rst]: maybe:
```diff
- The api-method provides environment-variables configuration:
+ ``ftw.slacker`` supports configuration through environment-variables:
```

File: README.rst [hunk 9]
```diff
@@ -1,21 +1,278 @@
...
+  from ftw.slacker import notify_slack
+
+  notify_slack(text="my first post")
 
-Add ``ftw.slacker`` to the list of eggs in your buildout, run buildout and
-restart your instance.
+Override the environment variable
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you set the STANDARD_SLACK_WEBHOOK environment variable, you can still use your own
```
Comment (jone) [inline README.rst]: `use your own` => `use a different`

File: README.rst [hunk 10]
```diff
@@ -1,21 +1,278 @@
...
+slack webhoock.
```
Comment (jone) [inline README.rst]: `webhoock` => `webhook`

File: README.rst [hunk 11]
```diff
@@ -1,21 +1,278 @@
...
+  export STANDARD_SLACK_WEBHOOK='https://hooks.slack.com/services/xxx'
+
+Each application will post messages to this slack webhook.
+
+Blacklist
+~~~~~~~~~
+
+Now you install a test-deployment on the same server where you want to deactivate the notifiactions.
+
+For this purpose, you can set another environment variable in this specific deployments builodut.cfg to
```
Comment (jone) [inline README.rst]: `deployments` => `deployment's`
`builodut.cfg` => ` ``buildout.cfg`` `

File: README.rst [hunk 12]
```diff
@@ -1,21 +1,278 @@
...
+the static NOTIFICATION_DEACTIVATION_VALUE:
```
Comment (jone) [inline README.rst]: This sounds as `NOTIFICATION_DEACTIVATION_VALUE` is the value; I'd maybe reword the sentence a bit 🤔

File: README.rst [hunk 13]
```diff
@@ -1,21 +1,278 @@
...
+
+This will reset the DEACTIVATE_SLACK_NOTIFICATION variable to `''`
+
+Deactivate through webhook_url
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's also possible to deactivate just a specific notification.
+
+Let's say, you have set your STANDARD_SLACK_WEBHOOK. We've learned, that
+you can call ``notify_slack`` without any webhoock_url to push notification
```
Comment (jone) [inline README.rst]: `webhoock_url` => `webhook_url`

File: README.rst [hunk 14]
```diff
@@ -1,21 +1,278 @@
...
+to the standard webhook url or you can call it with a more specific webhook url
+``notify_slack(webhook_url="xxx")`` to override the standard slack webhook.
+
+Using the NOTIFICATION_DEACTIVATION_VALUE as the webhook_url will just deactivate
+the current notification and will not bubble up to the standard slack webhook.
+
+This feature is excpecially useful for handling notifiaction comming from multiple
```
Comment (jone) [inline README.rst]: `excpecially` => `especially`
`notifiaction` => `notifiaction`

File: README.rst [hunk 15]
```diff
@@ -1,21 +1,278 @@
...
+external modules using the slacker-integration.
+
+See the next chapter for more information about advanced usage.
+
+Advance usage
+-------------
+
+Perhaps you get different external modules using the ``ftw.slacker`` implementation and
```
Comment (jone) [inline README.rst]: `you get different` 
-> `you've got different` or
-> `you use different`

File: README.rst [hunk 16]
```diff
@@ -1,21 +1,278 @@
...
+.. code:: python
+
+  from ftw.slacker import notify_slack
+  import os
+
+  def notify_user_login(user):
+      notify_slack(os.environ.get('FTW_LOGGER_SLACK_WEBHOOK'),
+                   text='User {} logged in'.format(user.username))
+
+If you don't set the `FTW_LOGGER_SLACK_WEBHOOK`, ``ftw.slacker`` will post the user
```
Comment (jone) [inline README.rst]: `` `FTW_LOGGER_SLACK_WEBHOOK` `` => I'd always use double ticks (` `` `) because pypi displays it better. I believe that the rest standard would also suggest using double ticks for such content.

File: README.rst [hunk 17]
```diff
@@ -1,21 +1,278 @@
...
+login to the default channel. If you set `FTW_LOGGER_SLACK_WEBHOOK`, ``ftw.slacker`` will
+use this more specific channel for notifications.
+
+Deactivating the whole notification system through the DEACTIVATE_SLACK_NOTIFICATION
+environment variable is not desired, because you still want to post other notifications,
+i.e. from your application which uses the standard slack webhook url.
+
+For this puropose, you can use the same NOTIFICATION_DEACTIVATION_VALUE for the more specific
```
Comment (jone) [inline README.rst]: Hmm, I think `NOTIFICATION_DEACTIVATION_VALUE` should be a specific string? It is not explained what that is.

File: README.rst [hunk 18]
```diff
@@ -1,21 +1,278 @@
...
+slack webhooks to deactivate just this specific notifiaction branch.
```
Comment (jone) [inline README.rst]: `notifiaction` 😏

File: README.rst [hunk 19]
```diff
@@ -1,21 +1,278 @@
...
+
+  [instance]
+  environment-vars +=
+      STANDARD_SLACK_WEBHOOK https://hooks.slack.com/services/xxx
+      FTW_LOGGER_SLACK_WEBHOOK deactivate
+
+Threading
+---------
+
+All requests to the Slack-API will be handled within its own threads.
```
Comment (jone) [inline README.rst:267]: Maybe:
> All messages are sent in a separate thread so that it is non-blocking and does not crash the application on an error. That's why the `notify_slack` function does not return a success state or raise an exception on errors.

File: ftw/slacker/interfaces.py [hunk 20]
```diff
@@ -0,0 +1,14 @@
+from zope.interface import Interface
+
+
+class ISlackNotifier(Interface):
+    """An ISlackNotifier utility is able to post messages through webhooks
+    into a slack channel.
+    """
+
+    def notify():
+        """This method perfomrs the slack-notification.
```
Comment (jone) [inline ftw/slacker/interfaces.py]: `perfomrs` => `performs`

File: ftw/slacker/interfaces.py [hunk 21]
```diff
@@ -0,0 +1,14 @@
+from zope.interface import Interface
+
+
+class ISlackNotifier(Interface):
+    """An ISlackNotifier utility is able to post messages through webhooks
+    into a slack channel.
+    """
+
+    def notify():
```
Comment (jone) [inline ftw/slacker/interfaces.py]: I'd declare the arguments here and then verify it in a test with `zope.interface.verify.verify[Class|Object]`.

File: ftw/slacker/slack_notifier.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
+from ftw.slacker.interfaces import ISlackNotifier
+from threading import Thread
+from zope.component import getUtility
+from zope.interface import implements
+import os
+import requests
+
+
+def notify_slack(*args, **kwargs):
+    """This is the main api mehtod to perform a slack notification
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: `mehtod` => `function`

File: ftw/slacker/slack_notifier.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+    Always use this method to do slack notifications.
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: `method` => `function`

File: ftw/slacker/slack_notifier.py [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+    # Name of the request thread
+    THREAD_NAME = 'SlackNotifier-Thread'
+
+    # Use this string in your environment variables to deactivate
+    # notification.
+    NOTIFICATION_DEACTIVATION_VALUE = 'deactivate'
+
+    # Possible environment variables
+    STANDARD_SLACK_WEBHOOK = 'STANDARD_SLACK_WEBHOOK'
+    DEACTIVATE_SLACK_NOTIFICATION = 'DEACTIVATE_SLACK_NOTIFICATION'
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: Why not use the strings directly?  As far as I see those are only used once, each.

File: ftw/slacker/slack_notifier.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+    def notify(self, webhook_url=None, timeout=2, verify=True, **payload):
+        """Performs the slack notification
+        """
+        if self._is_notifiaction_globally_deactivated():
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: typo `notifiaction`

File: ftw/slacker/slack_notifier.py [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+        thread.start()
+        return thread
+
+    def _do_request(self, webhook_url=None, timeout=2, verify=True, **payload):
+        """Actually performs the request.
+        """
+        requests.post(webhook_url,
+                      timeout=timeout,
+                      verify=verify,
+                      json=payload,).raise_for_status()
```
Comment (jone) [inline ftw/slacker/slack_notifier.py:62]: What happens when `raise_for_status` actually raises an exception?
Since we are in another thread there is nobody there to handle the exception.
Is it logged?
Did you test that?

File: ftw/slacker/slack_notifier.py [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+    a webhook.
+    """
+    implements(ISlackNotifier)
+
+    # Name of the request thread
+    THREAD_NAME = 'SlackNotifier-Thread'
+
+    # Use this string in your environment variables to deactivate
+    # notification.
+    NOTIFICATION_DEACTIVATION_VALUE = 'deactivate'
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: I'd probably move this to module-level.
That also has the advantage that it can be imported from outside.

File: ftw/slacker/slack_notifier.py [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+        an environment variable.
+        """
+        value = self._get_env_value(self.DEACTIVATE_SLACK_NOTIFICATION)
+        if not value:
+            return False
+
+        return value.lower() == self.NOTIFICATION_DEACTIVATION_VALUE
+
+    def _get_env_value(self, key):
+        return os.environ.get(key)
```
Comment (jone) [inline ftw/slacker/slack_notifier.py]: Why not just use `os.environ.get`?

File: README.rst (hunk 5 above)
Comment (jone) [reply README.rst:83]: I realized that the thread is returned. This is quite handy for testing 🤔

File: ftw/slacker/tests/__init__.py [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+    def grant(self, *roles):
+        setRoles(self.portal, TEST_USER_ID, list(roles))
+        transaction.commit()
+
+
+class ResponseStub(object):
+
+    def __init__(self, status_code):
+        self.status_code = status_code
```
Comment (jone) [inline ftw/slacker/tests/__init__.py]: I think you are not using `status_code`; could be removed from `ResponseStub ` and `RequestsMock` 🤔

**Review by jone — CHANGES_REQUESTED** [review]
Nice work 👍 🎉 
I've added some comments with opinions, typos and ideas 😎

File: ftw/slacker/tests/__init__.py (hunk 29 above)
Comment (elioschmutz) [reply ftw/slacker/tests/__init__.py]: 👍

File: ftw/slacker/slack_notifier.py (hunk 28 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: Perhaps we want to add more logic while getting the values somewhen in the future.

File: ftw/slacker/slack_notifier.py (hunk 26 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py:62]: Jeas, it is. The main thread will not be blocked, but the console will raise the error and tell the user, that the error occured in a separate thread:

bildschirmfoto 2017-07-26 um 15 01 43

I have no tests for this because it's very hard to test and the value is very low.

File: ftw/slacker/slack_notifier.py (hunk 25 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: 👍

File: ftw/slacker/slack_notifier.py (hunk 24 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: Because I want to use them in tests and perhaps in other implementations...

File: ftw/slacker/slack_notifier.py (hunk 27 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: 👍

File: ftw/slacker/slack_notifier.py (hunk 23 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: 👍

File: ftw/slacker/slack_notifier.py (hunk 22 above)
Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: 👍

File: ftw/slacker/interfaces.py (hunk 20 above)
Comment (elioschmutz) [reply ftw/slacker/interfaces.py]: 👍

File: README.rst (hunk 19 above)
Comment (elioschmutz) [reply README.rst:267]: 👍

File: README.rst (hunk 18 above)
Comment (elioschmutz) [reply README.rst]: 😠  whats wrong with my keyboard... 😇

File: README.rst (hunk 17 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 16 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 15 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 14 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 13 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 12 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 11 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 10 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 9 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 8 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 7 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 6 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 5 above)
Comment (elioschmutz) [reply README.rst:83]: 👍

File: README.rst (hunk 4 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 3 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 2 above)
Comment (elioschmutz) [reply README.rst]: 👍

File: README.rst (hunk 1 above)
Comment (elioschmutz) [reply README.rst]: 👍

Comment (elioschmutz) [reply README.rst]: 👍

File: ftw/slacker/interfaces.py (hunk 21 above)
Comment (elioschmutz) [reply ftw/slacker/interfaces.py]: 👍

**Comment by elioschmutz** [comment]
@jone thanks for your review! Could you take another look on it? I've made all changes within a new commit.

File: ftw/slacker/slack_notifier.py (hunk 26 above)
Comment (jone) [reply ftw/slacker/slack_notifier.py:62]: Cool 👍

File: ftw/slacker/slack_notifier.py (hunk 28 above)
Comment (jone) [reply ftw/slacker/slack_notifier.py]: Hmm, we shouldn't build for a potential, unplanned future. I think such extractions such be done when it actually brings value. It's more confusing for me as a reader, because I don't know what `_get_env_value` actually does. As pythonist I know what `os.environ.get` does.

Comment (elioschmutz) [reply ftw/slacker/slack_notifier.py]: 👍  thats true! I updated the code.

File: ftw/slacker/interfaces.py [hunk 30]
```diff
@@ -0,0 +1,21 @@
...
+class ISlackNotifier(Interface):
+    """An ISlackNotifier utility is able to post messages through webhooks
+    into a slack channel.
+    """
+
+    def notify(webhook_url, timeout, verify, **payload):
+        """This method performs the slack-notification.
+
+        @param webhook_url: The Slack webhook url
+        @param timeout: Raises a timeout error after the given seconds.
```
Comment (jone) [inline ftw/slacker/interfaces.py]: What kind of notation is that?
I have started using the [sphinx autodoc syntax](http://www.sphinx-doc.org/en/stable/ext/autodoc.html), for example in [ftw.publisher](https://github.com/4teamwork/ftw.testbrowser/blob/master/ftw/testbrowser/core.py#L210).
If this syntax has no deeper meaning I propose to also use the autodoc syntax, mainly for being a good example and good practice. I expect that we will start using that in other places too and it would be grat to have a consistent syntax 😉

**Review by jone — APPROVED** [review]
Awesome, thank you 🎉 ✨
